theano/tensor/opt.py

```diff
@@ -221,89 +221,124 @@ def inplace_elemwise_optimizer(fgraph):
+
+                verbose = False
+
+                raised_warning = not verbose
+
+                for candidate_output in candidate_outputs:
+                    for candidate_input in candidate_inputs:
+                        # remove inputs that don't have the same dtype as the
+                        # output
+                        if node.inputs[candidate_input].type != node.outputs[
+                                candidate_output].type:
+                            continue
+
+                        inplace_pattern = dict(baseline)
+                        inplace_pattern[candidate_output] = candidate_input
+                        try:
+                            if hasattr(op.scalar_op, "make_new_inplace"):
+                                new_scal = op.scalar_op.make_new_inplace(
+                                    scalar.transfer_type(
+                                        *[inplace_pattern.get(i, None) \
+                                              for i in xrange(len(node.outputs))]))
+                            else:
+                                new_scal = op.scalar_op.__class__(
+                                    scalar.transfer_type(
+                                        *[inplace_pattern.get(i, None) \
+                                              for i in xrange(len(node.outputs))]))
+                            new_outputs = OP(new_scal, inplace_pattern)(
                                 *node.inputs, **dict(return_list=True))
-                        new_node = new_outputs[0].owner
-
-                        for r, new_r in zip(node.outputs, new_outputs):
-                            fgraph.replace(r, new_r,
-                                        reason="inplace_elemwise_optimizer")
-                        nb_change_no_validate += 1
-                        if nb_change_no_validate >= check_each_change:
-                            fgraph.validate()
-                            chk = fgraph.checkpoint()
-                            nb_change_no_validate = 0
-                    except (ValueError, TypeError, InconsistencyError) as e:
-                        if check_each_change != 1 and not raised_warning:
-                            print((
-                                    "Some inplace optimization was not "
-                                    "performed due to unexpected error:"), file=sys.stderr)
-                            print(e, file=sys.stderr)
-                            raised_warning = True
-                        fgraph.revert(chk)
-                        continue
-                    candidate_inputs.remove(candidate_input)
-                    node = new_node
-                    baseline = inplace_pattern
-                    break
+                            new_node = new_outputs[0].owner
+
+                            # Add changed node into changed_node_since_validate
+                            if node not in changed_node_since_validate:
```

Comment:
Instead of doing all those check on a list, (that is costly) using a set would be better I think for changed_node_since_validate.